ceph_deploy/gatherkeys.py

```diff
@@ -1,84 +1,254 @@
-    return False
+def gatherkeys_missing(args, distro, rlogger, keypath, keytype, dest_dir):
+    """
+    Get or create the keyring from the mon using the mon keyring by keytype and
+    copy to dest_dir
+    """
+    arguments = [
+        '/usr/bin/ceph',
+        '--connect-timeout=25',
+        '--cluster={cluster}'.format(
+            cluster=args.cluster),
+        '--name', 'mon.',
+        '--keyring={keypath}'.format(
+            keypath=keypath),
+        'auth', 'get-or-create',
+        ]
+    identity = keytype_identity(keytype)
+    if identity is None:
+        raise RuntimeError('Could not find identity for keytype:%s' % keytype)
+    arguments.append(identity)
+    capabilites = keytype_capabilities(keytype)
+    if capabilites is None:
+        raise RuntimeError('Could not find capabilites for keytype:%s' % keytype)
+    arguments.extend(capabilites)
+    out, err, code = remoto.process.check(
+        distro.conn,
+        arguments
+        )
+    if code != 0:
+        rlogger.error('"ceph auth get-or-create for keytype %s returned %s' % (keytype, code))
+        for line in err:
+            rlogger.debug(line)
+        return False
+    keyring_name_local = keytype_path_to(args, keytype)
+    keyring_path_local = os.path.join(dest_dir, keyring_name_local)
+    with file(keyring_path_local, 'w') as f:
+        for line in out:
+            f.write(line + '\n')
+    return True
+
+
+def gatherkeys_with_mon(args, host, dest_dir):
+    """
+    Connect to mon and gather keys if mon is in quorum.
+    """
+    distro = hosts.get(host, username=args.username)
+    dir_keytype_mon = ceph_deploy.util.paths.mon.path(args.cluster, host)
+    path_keytype_mon = "%s/keyring" % (dir_keytype_mon)
+    mon_key = distro.conn.remote_module.get_file(path_keytype_mon)
+    if mon_key is None:
+        LOG.warning("No mon key found. Is '%s' a mon node" % (host))
+        return False
+    mon_name_local = keytype_path_to(args, "mon")
+    mon_path_local = os.path.join(dest_dir, mon_name_local)
+    with file(mon_path_local, 'w') as f:
+        f.write(mon_key)
+    rlogger = logging.getLogger(host)
+    path_asok = ceph_deploy.util.paths.mon.asok(args.cluster, host)
+    out, err, code = remoto.process.check(
+        distro.conn,
+            [
+                "/usr/bin/ceph",
+                "--connect-timeout=25",
+                "--cluster={cluster}".format(
+                    cluster=args.cluster),
+                "--admin-daemon={asok}".format(
+                    asok=path_asok),
+                "mon_status"
+            ]
+        )
+    if code != 0:
+        rlogger.error('"ceph mon_status %s" returned %s' % (host, code))
+        for line in err:
+            rlogger.debug(line)
+        return False
+    try:
+        mon_status = json.loads("".join(out))
+    except ValueError:
+        rlogger.error('"ceph mon_status %s" output was not json' % (host))
```

Comment:
at this point it isn't clear what is it that `ceph mon_status` got that wasn't JSON. Maybe log the output as part of the error?